File: Python/algorithmic_toolbox/07_huge_fibonacci_modulo.py
import sys


def convert_to_intlist(line):
    return [int(elem) for elem in line.split()]


def calc_pisano_period(modulus):
    p = 1
    a, b = 1, 1
    while not (a == 0 and b == 1):
        a, b = b, ((a + b) % modulus)
        p += 1
    return p


def calc_fibonacci_modulo(n, modulus):
    k = 1
    a, b = 0, 1
    while k <= n:
        a, b = b, (a + b) % modulus
        k += 1
    return a


def calc_huge_fibonacci_modulo(n, modulus):
    p = calc_pisano_period(modulus)
    return calc_fibonacci_modulo(n % p, modulus)


def main():
    data = sys.stdin.read().splitlines()
    n, modulus = convert_to_intlist(data[0])
    f = calc_huge_fibonacci_modulo(n, modulus)
    print(f)


if __name__ == "__main__":
    main()

File: R/algorithmic_toolbox/07_huge_fibonacci_modulo.R
#!/usr/bin/env Rscript
CalcPisanoPeriod <- function(modulus) {
    p <- 1
    a <- 1
    b <- 1
    while (!(a == 0 && b == 1)) {
        c <- (a + b) %% modulus
        a <- b
        b <- c
        p <- p + 1
    }
    p
}

CalcFibonacciModulo <- function(n, modulus) {
    k <- 1
    a <- 0
    b <- 1
    while (k <= n) {
        c <- (a + b) %% modulus
        a <- b
        b <- c
        k <- k + 1
    }
    a
}

CalcHugeFibonacciModulo <- function(n, modulus) {
    p <- CalcPisanoPeriod(modulus)
    CalcFibonacciModulo(n %% p, modulus)
}

Main <- function() {
    conn <- file("stdin", "r")
    line <- readLines(conn, n = 1)
    lst <- as.integer(unlist(strsplit(line, split = " ")))
    close(conn)
    f <- CalcHugeFibonacciModulo(lst[1], lst[2])
    cat(f, fill = TRUE)
}

if (!interactive()) {
    Main()
}
